walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,132 @@
+
+    if (attendees.isEmpty() && optionalAttendees.isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }else 
+    if (duration > TimeRange.WHOLE_DAY.duration()){
+      return Arrays.asList();
+    }
+
+    List<TimeRange> takenTime = new ArrayList<TimeRange>();
+    List<TimeRange> optionalBusy = new ArrayList<TimeRange>();
+    setTakenTimes(events, takenTime, optionalBusy, attendees, optionalAttendees);
+
+    List<TimeRange> mandatoryFree = new ArrayList<TimeRange>();
+    mandatoryFree.add(TimeRange.WHOLE_DAY);
+    List<TimeRange> optionalFree = new ArrayList<TimeRange>();
+    setFreeTimes(mandatoryFree, takenTime, duration);
+
+    if(takenTime.isEmpty()){
+      optionalFree.add(TimeRange.WHOLE_DAY);
+      setFreeTimes(optionalFree, optionalBusy, duration);
+      return optionalFree;
+    }else {
+      for(TimeRange t: mandatoryFree){
+        checkOptional(optionalFree, optionalBusy, t);
+      }
+    }
+
+    if(!optionalFree.isEmpty()){
+      return optionalFree;
+    } else {
+      return mandatoryFree;
+    } 
+  }
+
+  private static void setTakenTimes(Collection<Event> events, List<TimeRange> takenTime, List<TimeRange> optionalBusy,
+      Collection<String> attendees, Collection<String> optionalAttendees){
+
+    for(Event e: events){
+      Set<String> mandatoryAtt = new HashSet<>();
+      Set<String> optionalAtt = new HashSet<>();
+      Set<String> eventAtt = e.getAttendees();
+
+      mandatoryAtt.addAll(eventAtt);
+      mandatoryAtt.retainAll(attendees);
+      optionalAtt.addAll(eventAtt);
+      optionalAtt.retainAll(optionalAttendees);
+
+      if(mandatoryAtt.isEmpty()){
+        if(!optionalAtt.isEmpty()){
+        optionalBusy.add(e.getWhen());
+      }
+        continue;
+      }
+      takenTime.add(e.getWhen());
+    }
+    Collections.sort(takenTime, TimeRange.ORDER_BY_START);
+  }
+
+  private static void setFreeTimes(List<TimeRange> mandatoryFree, List<TimeRange> takenTime, long duration ){
+    for(TimeRange t: takenTime){
+      int lastIndex = mandatoryFree.size()-1;
+      TimeRange latestRange = mandatoryFree.get(lastIndex);
+      int currentStart = latestRange.start();
+      int currentEnd = latestRange.end();
+
+      int takenStart = t.start();
+      int takenEnd = t.end();
+
+      if(currentStart == takenStart || currentStart > takenStart){
+        if(currentStart > takenEnd){
+          continue;
+        }
+        if(takenEnd < currentEnd){
+          setRange(mandatoryFree, duration, lastIndex, takenEnd, currentEnd, false);
+        }else{
+          mandatoryFree.remove(lastIndex);
+        }
+      }else {
+        setRange(mandatoryFree, duration, lastIndex, currentStart, takenStart, false); 
+        if(currentEnd > takenEnd){
+          addRange(mandatoryFree, duration, takenEnd, currentEnd, false);
+        }
+      }
+    }
+  }
+
+  private static void setRange(List<TimeRange> freeTime, long duration, int index, int start, int end, boolean cond){
```

Comment:
As I explained passing around duration was a design choice to save time, but I can implement the filter as you can explained without noticeably compromising speed.

And I just pass false to the constructor as recommended, because I never need to use true.